### PR TITLE
Forked from ivan-kostko/go-sandbox

### PR DESCRIPTION
**Pull request by ivan-kostko**
Forked starting from: https://github.com/ivan-kostko/go-sandbox/commit/c944b71c48e4dfe4a900dcf44a2c08ce6ca9e12d

### CODE REVIEW

File: WorkerPool/WorkerPool.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	cancellationChan chan struct{}
+}
+
+// A new Deapertment Factory
+func NewWorkerPool(initWorkerNumber int) WorkerPool {
+	// instantiate  pool
+	workersChan := make(chan struct{}, initWorkerNumber)
+
+	// fill up pool
+	// for each initially empty slot we shoul put one value
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go]: typo: `should`

File: WorkerPool/WorkerPool_test.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+			for i := 0; i < startWorkerNumber; i++ {
+				dep.Do(func() { started <- struct{}{}; <-block }, timeOut)
+			}
+
+			// Sleep for all timeouts
+			time.Sleep(timeOut * time.Duration(initWorkerNumber))
+
+			actualStartedWorkers := actualWorkerNumber
+
+			// cancell workers
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: typo: `cancel`

**Review by kobergj — APPROVED** [review]
Looks awesome @ivan-kostko. Can't wait to work with it

**Comment by ivan-kostko** [comment]
@kobergj pls CR fixes

File: WorkerPool/WorkerPool.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
+package WorkerPool
+
+import (
+	"errors"
+	"time"
+)
+
+const (
+	ERR_WORKERPOOLSHUTDOWN = "The worker pool is shutting down and wont take new assignments"
+	ERR_TIMEDOUTREQUSTSLOT = "The request for a free execution slot has been timed out"
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go]: I guess you mean `ERR_TIMEOUTREQUESTSLOT`, don't you?

File: WorkerPool/WorkerPool.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+)
+
+// Represents simple assignment.
+type WorkItem func()
+
+// Represents simple workers pool operating on Projects and WorkItems.
+type WorkerPool interface {
+
+	// Synchronously requests worker slot and end exectes/does WorkItem in parallel routine as soon as slot is obtained.
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go]: `and end executes`?

File: WorkerPool/WorkerPool.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	// If no slot aquired upon timeOut exceeds - returns ERR_TIMEDOUTREQUSTSLOT.
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go]: Typos: `acquired` , `ERR_TIMEOUTREQUESTSLOT`

File: WorkerPool/WorkerPool.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	// If worker pool is already closed or closing while obtaining worker slot - returns ERR_WORKERPOOLSHUTDOWN.
+	Do(wi WorkItem, timeOut time.Duration) error
+
+	// Closes the worker pool.
+	// All new requests will be rejected returning ERR_WORKERPOOLSHUTDOWN.
+	// All requests waiting for slots should be notified and canceled returning ERR_WORKERPOOLSHUTDOWN.
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go]: typo: `cancelled`

File: WorkerPool/WorkerPool.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+	return &workerPool{
+		workersChan:      workersChan,
+		cancellationChan: cancellationChan,
+	}
+
+}
+
+// Implements WorkerPool.Do(wi WorkItem) method
+func (this *workerPool) Do(wi WorkItem, timeOut time.Duration) error {
```
Comment (kobergj) [inline WorkerPool/WorkerPool.go:59]: I think, due to consistency, we should also stick to CustomErrors here...

File: WorkerPool/WorkerPool_test.go [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+
+	for _, testCase := range testCases {
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedStartedWorkers := testCase.ExpectedStartedWorkers
+
+		fn := func(t *testing.T) {
+
+			// chanel workers are waiting for close
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: `chanel workers`? Isn't that a bit girly? I think `channel workers` would be a better name

File: WorkerPool/WorkerPool_test.go [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+			StartWorkerNumber:    1000,
+			ExpectedDonedWorkers: 1000,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedDoneWorkers := testCase.ExpectedDonedWorkers
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: Is it `expectedDONEworkers` or `expectedDONEDworkers`?

File: WorkerPool/WorkerPool_test.go [hunk 10]
```diff
@@ -0,0 +1,678 @@
...
+			for i := 0; i < startWorkerNumber; i++ {
+				wg.Add(1)
+				err := dep.Do(func() {
+					mtx.Lock()
+					actualWorkerDone++
+					mtx.Unlock()
+					wg.Done()
+				}, timeOut)
+				if err != nil {
+					wg.Done()
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go:216]: Isn't it decrementing twice here? Once on the end of `Do` func and then again if no Error occured?

File: WorkerPool/WorkerPool_test.go [hunk 11]
```diff
@@ -0,0 +1,678 @@
...
+		ExpectedWorkerChanLen           int
+		ExpectedIsCancelationChanClosed bool
+		ExpectedIsWorkersChanClosed     bool
+	}{
+		{
+			TestAlias:                       "20 work items for 0 workers",
+			InitWorkerNumber:                0,
+			StartWorkerNumber:               20,
+			ExpectedWorkerChanLen:           0,
+			ExpectedIsCancelationChanClosed: true,
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: typo `...cancellation...`

File: WorkerPool/WorkerPool_test.go [hunk 12]
```diff
@@ -0,0 +1,678 @@
...
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedWorkerChanLen := testCase.ExpectedWorkerChanLen
+		expectedIsCancelationChanClosed := testCase.ExpectedIsCancelationChanClosed
+		expectedIsWorkersChanClosed := testCase.ExpectedIsWorkersChanClosed
+
+		fn := func(t *testing.T) {
+
+			// chanel workers are waiting for close
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: You really like `chanel`, right?

File: WorkerPool/WorkerPool_test.go [hunk 13]
```diff
@@ -0,0 +1,678 @@
...
+			wg := sync.WaitGroup{}
+
+			for i := 0; i < startWorkerNumber; i++ {
+				wg.Add(1)
+				go func() { wg.Done(); dep.Do(func() { <-block }, 10) }()
+			}
+
+			wg.Wait()
+
+			// chanel closed on Close() complete
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: ...

File: WorkerPool/WorkerPool_test.go [hunk 14]
```diff
@@ -0,0 +1,678 @@
...
+			<-closed
+
+			wp := dep.(*workerPool)
+
+			actualWorkerChanLen := len(wp.workersChan)
+
+			var actualIsCancelationChanClosed bool
+
+			if _, more := <-wp.cancellationChan; !more {
+				actualIsCancelationChanClosed = true
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: Also: `...cancellation...`

File: WorkerPool/WorkerPool_test.go [hunk 15]
```diff
@@ -0,0 +1,678 @@
...
+			if _, more := <-wp.workersChan; !more {
+				actualIsWorkersChanClosed = true
+			}
+
+			if actualWorkerChanLen != expectedWorkerChanLen {
+				t.Errorf("For TestAlias '%s' WorkerPool.Close() with immidiate release of workers \r\n has WorkerChanLen length %v workers \r\n while expected %v \r\n", testAlias, actualWorkerChanLen, expectedWorkerChanLen)
+			}
+
+			if actualIsCancelationChanClosed != expectedIsCancelationChanClosed {
+				t.Errorf("For TestAlias '%s' WorkerPool.Close() with immidiate release of workers \r\n has CancelationChan Closed ( %v )  \r\n while expected %v \r\n", testAlias, actualIsCancelationChanClosed, expectedIsCancelationChanClosed)
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: typo? `immediate`? Also in other two `t.Errorf`

File: WorkerPool/WorkerPool_test.go [hunk 16]
```diff
@@ -0,0 +1,678 @@
...
+
+	for _, testCase := range testCases {
+		testAlias := testCase.TestAlias
+		initWorkerNumber := testCase.InitWorkerNumber
+		startWorkerNumber := testCase.StartWorkerNumber
+		expectedDoneWorkers := testCase.ExpectedDoneWorkers
+
+		fn := func(t *testing.T) {
+
+			// chanel workers are waiting for close
```
Comment (kobergj) [inline WorkerPool/WorkerPool_test.go]: :expressionless:

**Review by kobergj — CHANGES_REQUESTED** [review]
неплохо @ivan-kostko, but please FOR my comments. (and fix typos!!)

File: WorkerPool/WorkerPool_test.go (hunk 10 above)
Comment (ivan-kostko) [reply WorkerPool/WorkerPool_test.go:216]: Not really. `wg.Done()` is invoked whether `func()` is executed, or if it is not and any error returned.

Comment (ivan-kostko) [reply WorkerPool/WorkerPool_test.go:216]: To be 100%% clear, it should be wrapped into local anonymous function and `defer wg.Done()`. However, actually there is a bug in testing package, which invokes `defer` in a very specific way.

File: WorkerPool/WorkerPool.go (hunk 7 above)
Comment (ivan-kostko) [reply WorkerPool/WorkerPool.go:59]: CustomError implements `error` interface.

**Pull request merged**